### DISCUSSION TITLE
Scratch license change

### DISCUSSION THREAD

**Discussion by mxmou** [ScratchAddons/ScratchAddons#7958] [Discussion]
Scratch just changed the license for all their code to AGPL: https://www.scratchfoundation.org/open-source-license

**Comment by DNin01** [comment]
@WorldLanguages, will this affect Scratch Addons? Will we have to switch to AGPL too?

**Reply by WorldLanguages** [reply]
Honestly I'm not an expert.

**Reply by DNin01** [reply]
Maybe if we use icons or code snippets from updated versions of Scratch.

It might be a good idea going forward to try to fully switch to AGPL anyway, to follow suit with Scratch.

**Reply by GarboMuffin** [reply]
Not a lawyer, haven't talked to one

There are at least a few addons that interact so closely with Scratch's open source packages like scratch-gui that they are probably considered derivative works of Scratch. If this interpretation is accurate, then I believe the resulting addons would have to be under AGPL-3.0 if they were written or updated today. My understanding of copyright law could also be completely wrong.

Relevant legal text from section 13 of the GPL-3.0:

> Notwithstanding any other provision of this License, you have permission to link or combine any covered work with a work licensed under version 3 of the GNU Affero General Public License into a single combined work, and to convey the resulting work. The terms of this License will continue to apply to the part which is the covered work, but the special requirements of the GNU Affero General Public License, section 13, concerning interaction through a network will apply to the combination as such.

Seems like having a mix of AGPL-3.0 and GPL-3.0 in one project is perfectly fine (combined work is effectively AGPL-3.0 though individual components can't retroactively change license), so if you wanted to change all future Scratch Addons changes to be AGPL-3.0 you could do that. Doing so might avoid some ambiguity.

**Reply by DNin01** [reply]
Is it possible to convert Scratch Addons into an AGPL-3.0-licensed project without needing to ask anyone for permission or would our project remain a mix of GPL-3.0 and AGPL-3.0?

(Splitting the addons and the core components of Scratch Addons up into separate repositories might also make mixing licenses simpler.)

**Reply by WorldLanguages** [reply]
> There are at least a few addons that interact so closely with Scratch's open source packages like scratch-gui

Fortunately, most of those addons were programmed by a few contributors that are still around.

**Reply by GarboMuffin** [reply]
Not a lawyer, haven't talked to one

I believe you can say something along the lines "All commits made on or after January XX 2025 are licensed under AGPL-3.0. All prior commits are GPL-3.0." You wouldn't need to get anyone's permission to do this; the permission to combine AGPL-3.0 and GPL-3.0 works into one combined work is granted by section 13.

The extra requirements from AGPL-3.0 wouldn't apply to the parts that were already GPL-3.0. Anyone who cares about that difference can use git blame to figure out when the commits in question were made to determine what license applies.

**Reply by DNin01** [reply]
> Fortunately, most of those addons were programmed by a few contributors that are still around.

In that case, if we are interested in transitioning to AGPL-3.0, one thing I think we could do is say that our code is AGPL unless otherwise specified and then any code that is still partly or fully GPL could have a GPL boilerplate added.

Any code where the contributor allows us to could be converted to AGPL, if for any reason we would need to.

**Reply by WorldLanguages** [reply]
> I believe you can say something along the lines "All commits made on or after January XX 2025 are licensed under AGPL-3.0. All prior commits are GPL-3.0." You wouldn't need to get anyone's permission to do this; the permission to combine AGPL-3.0 and GPL-3.0 works into one combined work is granted by section 13.
\
The extra requirements from AGPL-3.0 wouldn't apply to the parts that were already GPL-3.0. Anyone who cares about that difference can use git blame to figure out when the commits in question were made to determine what license applies.

That would be ideal.

> our code is AGPL unless otherwise specified

I think that can get complicated quickly. As GarboMuffin said, it's better for us if those that actually want to make sure the license is ok for their usecase, review the situation themselves.

**Reply by GarboMuffin** [reply]
GPL-3.0 suggests putting a comment at the start of each file like

```
    <one line to give the program's name and a brief idea of what it does.>
    Copyright (C) <year>  <name of author>

    This program is free software: you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation, either version 3 of the License, or
    (at your option) any later version.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with this program.  If not, see <https://www.gnu.org/licenses/>.
```

If we were to do that, difference between GPL-3.0 and AGPL-3.0 components would be explicitly stated in the comment. Existing files would say "GNU General Public License" and new files would say "GNU Affero General Public License".

Personally, I'm not a big fan of the massive boilerplate, and I'm not sure how to update the comment if a GPL-3.0 file is modified with new code that wants to be under AGPL-3.0, so I wouldn't recommend this

**Comment by DNin01** [comment]
I wonder if they decided to change their license after they saw Gandi IDE. 😄

**Reply by WorldLanguages** [reply]
No way to know.

The license choice for this project was quite arbitrary as well, but well, I was a teenager and didn't expect we'd get 80+ contributors.

**Reply by Secret-chest** [reply]
Maybe, and I think Gandi IDE should be forced to free it.
